concordium-consensus/src/Concordium/GlobalState/TreeState.hs

```diff
@@ -499,3 +504,114 @@ importBlockV2 blockBS tm logm logLvl continuation =
       logm logLvl LLError $ "Can't deserialize block: " ++ show err
       return SerializationFail
     Right block -> continuation block
+
+-- | Exists so we can have the `ProtocolVersion` in the `Reader` computation ctx. 
+newtype ProtocolVersionedReaderT (pv :: ProtocolVersion) r m a = ProtocolVersionedReaderT {runProtocolVersionedReaderT :: ReaderT r m a}
+  deriving (Functor, Applicative, Monad, MonadReader r, MonadTrans)
+
+instance BlockStateTypes (ProtocolVersionedReaderT pv r m) where
+    type BlockState (ProtocolVersionedReaderT pv r m) = BlockState m
+    type UpdatableBlockState (ProtocolVersionedReaderT pv r m) = UpdatableBlockState m
+    type Account (ProtocolVersionedReaderT pv r m) = Account m
+
+-- |A `BlockItem` with verification context.
+-- The verification `Context` is used for verifying the transaction.
+-- The transaction is only verified w
+-- result was `definitelyNotValid`.
+type BlockItemWithVerificationContext t = (BlockItem, Context (BlockState t))
+
+-- |Context required for carrying out transaction verification
+-- in the reader based instance.
+-- todo doc.
+data Context t = Context {
+  _ctxBs :: t,
+  _ctxMaxBlockEnergy :: !Energy,
+  -- |Whether the transaction was received from a block or individually.
+  _isTransactionFromBlock :: Bool
+  }
+makeLenses ''Context
+
+-- |Determines if a transaction definitely cannot be valid now or in a future block.
+-- Transactions received individually must be verified successfully.
+-- However transactions there is a looser requirement for transactions received
+-- as part of a block. See the 'TransactionVerification' module for more details.
+definitelyNotValid :: TVer.VerificationResult -> Bool -> Bool
+definitelyNotValid verificationResult fromBlock =
+  case fromBlock of
+    False ->
+      case verificationResult of
+        TVer.Ok _ -> False
+        TVer.MaybeOk _ -> True
+        TVer.NotOk _ -> True
+    True ->
+      case verificationResult of
+        TVer.NotOk _ -> True
+        TVer.Ok _ -> False
+        TVer.MaybeOk _ -> False
```

Comment:
I moved  `definitelyNotValid` back to the `TransactionVerification` module.
 
I'm a bit in doubt with the others. The reader based `TransactionVerifier`, the `Context` and `ProtocolVersionedReaderT` since they are only used within the `TreeState` and the reader based `TransactionVerifier` has a constraint on the `TreeState` so it would require some more refactoring. 

Not saying I don't want to move the definitions - but I'm not sure it will end up in a better solution.